### PR TITLE
test: support .env.test to reuse env variables across the tests 

### PR DESCRIPTION
**Pull request by nazarhussain**
**Motivation**

Reuse the env variables. 

**Description**

- Single source for env variables during testing `.env.test`
- Load same file in simulation tests
- Load same file in the other tests

**Steps to test or reproduce**

- Run all tests

### CODE REVIEW

File: packages/cli/package.json [hunk 1]
```diff
@@ -32,11 +32,11 @@
     "lint:fix": "yarn run lint --fix",
     "test:unit": "vitest --run --dir test/unit/",
     "test:e2e": "vitest --run --config vitest.e2e.config.ts --dir test/e2e/",
-    "test:sim:multifork": "LODESTAR_PRESET=minimal node --loader ts-node/esm test/sim/multi_fork.test.ts",
-    "test:sim:mixedclient": "LODESTAR_PRESET=minimal node --loader ts-node/esm test/sim/mixed_client.test.ts",
-    "test:sim:endpoints": "LODESTAR_PRESET=minimal node --loader ts-node/esm test/sim/endpoints.test.ts",
-    "test:sim:deneb": "LODESTAR_PRESET=minimal node --loader ts-node/esm test/sim/deneb.test.ts",
-    "test:sim:backup_eth_provider": "LODESTAR_PRESET=minimal node --loader ts-node/esm test/sim/backup_eth_provider.test.ts",
+    "test:sim:multifork": "DOTENV_CONFIG_PATH=../../.env.test LODESTAR_PRESET=minimal node -r dotenv/config --loader ts-node/esm test/sim/multi_fork.test.ts",
```
Comment (matthewkeil) [inline packages/cli/package.json]: should LODESTAR_PRESET go into the .env?

File: vitest.base.unit.config.ts [hunk 2]
```diff
@@ -1,6 +1,6 @@
 import path from "node:path";
 import {defineConfig} from "vitest/config";
-const __dirname = new URL(".", import.meta.url).pathname;
+const currentDir = new URL(".", import.meta.url).pathname;
```
Comment (matthewkeil) [inline vitest.base.unit.config.ts]: I’m partial to leaving this as `__dirname`.  I dont feel strongly about it but it’s sort of the “node way” to name that

**Review by matthewkeil — CHANGES_REQUESTED** [review]
Such a good idea!!!

**Review by nflaig — COMMENTED** [review]
It a bit counter intuitive to me to have .env files in source control. I am rather used to having those files to override locally and explicitly not have them tracked by git.

Why can't we just set the default image versions in `scripts/run_e2e_env.sh` and allow to override them via env variables?

We could consider having dotenv module to override those but seems a bit overkill as there are not many envs and you can just add `GETH_DOCKER_IMAGE=...` in front of the yarn command to achieve this easily if someone wants to use a different version than the default one.

**Comment by nazarhussain** [comment]
> It a bit counter intuitive to me to have .env files in source control

I am also not in favor of keeping `.env` file in version control as it's used to pass some sensitive data to application and many tooling use it passively so you can't protect values to be transmitted somewhere. But here we don't have a `.env` file we have a `.env.test` which is a specific use case and we use that file explicitly where we needed. We could use any name e.g. `TestEnv.txt` does not matter. If we want to reuse definition of env variables around tests and workflows we need to define those somewhere. 

>  I am rather used to having those files to override locally and explicitly not have them tracked by git.

Not now but if needed we can make this setup in a way that you can override those variables in `.env.local` file and obviously that file will be git ignored.

File: packages/cli/package.json (hunk 1 above)
Comment (nazarhussain) [reply packages/cli/package.json]: Yes we can add that there as well. Need to check if any other test file depend on other preset.

File: vitest.base.unit.config.ts (hunk 2 above)
Comment (nazarhussain) [reply vitest.base.unit.config.ts]: Yes I had that initially as `__dirname` here and other places as well, but feel like a old pattern now.

**Comment by nflaig** [comment]
> But here we don't have a .env file we have a .env.test

~~Maybe that's just me, but I don't see a difference between those two files, neither should be in source control imo, e.g Nextjs also uses .env.test file for [Test Environment Variables](https://nextjs.org/docs/pages/building-your-application/configuring/environment-variables#test-environment-variables)~~

But I like the de-duplication of setting the version which is much better than before

**Edit:** It seems to be convention to have  .env.test as part of source control
image

So I guess we are good to go

File: .github/actions/dotenv/action.yml [hunk 3]
```diff
@@ -0,0 +1,5 @@
+name: "Setup env variables using .env file"
+description: "Load .env file from root of repo and setup for git runner"
```
Comment (nflaig) [inline .github/actions/dotenv/action.yml]: what is a "git runner"? I think this should either say "workflow runner", "CI runner" or just simply "runner"

File: vitest.base.unit.config.ts (hunk 2 above)
Comment (nflaig) [reply vitest.base.unit.config.ts]: For consistency with the rest of the codebase `__dirname` is better, but no strong opinion here

File: packages/cli/package.json (hunk 1 above)
Comment (matthewkeil) [reply packages/cli/package.json]: Ok.  Then let's definitely move that so we are using one pattern (ie and env file)

File: CONTRIBUTING.md [hunk 4]
```diff
@@ -66,7 +66,7 @@ If you observe following error running any of the test files that means you are
 
 ```sh
 cd packages/beacon-node
-LODESTAR_PRESET=minimal yarn vitest --run --bail 1 --config vitest.spec.config.ts test/spec/presets/sanity.test.ts -t attester_slashing
```
Comment (nflaig) [inline CONTRIBUTING.md:69]: just to confirm, if I run this command in a package directory (e.g. packages/beacon-node) will it load the environment variables in .env.test?

**Review by nflaig — DISMISSED** [review]

File: CONTRIBUTING.md (hunk 4 above)
Comment (nazarhussain) [reply CONTRIBUTING.md:69]: Yes

File: .env.test [hunk 5]
```diff
@@ -0,0 +1,8 @@
+# We use these images during sim and e2e tests
+GETH_DOCKER_IMAGE=ethereum/client-go:v1.13.11
+# Use either image or local binary for the testing
+GETH_BINARY_DIR=
+LIGHTHOUSE_DOCKER_IMAGE=sigp/lighthouse:v4.6.0-amd64-modern-dev
+NETHERMIND_DOCKER_IMAGE=nethermind/nethermind:1.25.3
+# We mostly use mainnet for unit testsing
```
Comment (nflaig) [inline .env.test]:
```suggestion
# We mostly use mainnet for unit testing
```

File: CONTRIBUTING.md [hunk 6]
```diff
@@ -63,10 +63,10 @@ If you observe following error running any of the test files that means you are
 - Spec tests often compare full expected vs actual states in JSON format.
 - A single logical error can cause many spec tests to fail. To focus on a single test at a time you can use vitest's option `--bail 1` to stop at the first failed test
 - To then run only that failed test you can run against a specific file as use vitest's filters option `-t <pattern>` to run only one case
+- Before running the tests, make sure to switch to the package directory (e.g. `packages/beacon-node`) to speed up test execution
 
 ```sh
-cd packages/beacon-node
-LODESTAR_PRESET=minimal yarn vitest --run --bail 1 --config vitest.spec.config.ts test/spec/presets/sanity.test.ts -t attester_slashing
```
Comment (nflaig) [inline CONTRIBUTING.md:69]: I think we should actually keep `LODESTAR_PRESET=minimal` as this demonstrates here how to run spec tests with different presets

File: packages/beacon-node/package.json [hunk 7]
```diff
@@ -76,10 +76,10 @@
     "lint": "eslint --color --ext .ts src/ test/",
     "lint:fix": "yarn run lint --fix",
     "test": "yarn test:unit && yarn test:e2e",
-    "test:unit:minimal": "vitest --run --segfaultRetry 3 --dir test/unit/",
-    "test:unit:mainnet": "LODESTAR_PRESET=mainnet vitest --run --dir test/unit-mainnet",
-    "test:unit": "wrapper() { yarn test:unit:minimal $@ && yarn test:unit:mainnet $@; }; wrapper",
-    "test:e2e": "LODESTAR_PRESET=minimal vitest --run --segfaultRetry 3 --config vitest.e2e.config.ts --dir test/e2e",
+    "test:unit:minimal": "LODESTAR_PRESET=minimal vitest --run --segfaultRetry 3 --dir test/unit/",
+    "test:unit:mainnet": "vitest --run --segfaultRetry 3 --dir test/unit-mainnet",
```
Comment (nflaig) [inline packages/beacon-node/package.json]: this command is mainnet preset specific, I don't think it is good that this can be executed with a different preset if the value in .env.test is changed

```suggestion
    "test:unit:mainnet": "LODESTAR_PRESET=mainnet vitest --run --segfaultRetry 3 --dir test/unit-mainnet",
```

File: packages/beacon-node/package.json [hunk 8]
```diff
@@ -76,10 +76,10 @@
     "lint": "eslint --color --ext .ts src/ test/",
     "lint:fix": "yarn run lint --fix",
     "test": "yarn test:unit && yarn test:e2e",
-    "test:unit:minimal": "vitest --run --segfaultRetry 3 --dir test/unit/",
-    "test:unit:mainnet": "LODESTAR_PRESET=mainnet vitest --run --dir test/unit-mainnet",
-    "test:unit": "wrapper() { yarn test:unit:minimal $@ && yarn test:unit:mainnet $@; }; wrapper",
```
Comment (nflaig) [inline packages/beacon-node/package.json:81]: This wrapper was just added recently by @jeluard to allow forwarding vitest flags like `-t` to all executed commands
- https://github.com/ChainSafe/lodestar/pull/6249

File: packages/beacon-node/package.json [hunk 9]
```diff
@@ -89,7 +89,7 @@
     "test:spec:bls": "vitest --run --config vitest.spec.config.ts --dir test/spec/bls/",
     "test:spec:general": "vitest --run --config vitest.spec.config.ts --dir test/spec/general/",
     "test:spec:minimal": "LODESTAR_PRESET=minimal vitest --run --config vitest.spec.config.ts --dir test/spec/presets/",
-    "test:spec:mainnet": "LODESTAR_PRESET=mainnet vitest --run --config vitest.spec.config.ts --dir test/spec/presets/",
+    "test:spec:mainnet": "vitest --run --config vitest.spec.config.ts --dir test/spec/presets/",
```
Comment (nflaig) [inline packages/beacon-node/package.json]: I don't get why `LODESTAR_PRESET=mainnet` is removed here, if these are specific to mainnet preset and I would expect them to run against that preset no matter what values i have in my .env files.

Same goes for any tests that only pass for a specific preset, we should not allow to override it there

Comment (nazarhussain) [reply packages/beacon-node/package.json]: Because `LODESTAR_PRESET=mainnet ` is passed as default form the `.env.test` file.

File: packages/beacon-node/package.json (hunk 7 above)
Comment (nazarhussain) [reply packages/beacon-node/package.json]: Don't need mainnet preset as it will be passed from the env file.

Comment (nflaig) [reply packages/beacon-node/package.json]: > Don't need mainnet preset as it will be passed from the env file.

But I don't get the purpose of the value set in .env file to be overriden by a user? I don't get the purpose of the .env file if it not that the user can easily override the value within

File: packages/logger/package.json [hunk 10]
```diff
@@ -61,7 +61,7 @@
     "test:browsers:chrome": "vitest --run --browser chrome --config ./vitest.browser.config.ts --dir test/unit",
     "test:browsers:firefox": "vitest --run --browser firefox --config ./vitest.browser.config.ts --dir test/unit",
     "test:browsers:electron": "echo 'Electron tests will be introduced back in the future as soon vitest supports electron.'",
-    "test:e2e": "LODESTAR_PRESET=minimal vitest --run --config vitest.e2e.config.ts --dir test/e2e",
```
Comment (nflaig) [inline packages/logger/package.json:64]: Before we would run those with minimal preset, is this change intended?

File: packages/params/test/e2e/overridePreset.test.ts [hunk 11]
```diff
@@ -21,8 +21,7 @@ describe("Override preset", function () {
   vi.setConfig({testTimeout: 30_000});
 
   it("Should correctly override preset", async () => {
-    // These commands can not run with minimal preset
-    if (process.env.LODESTAR_PRESET === "minimal") delete process.env.LODESTAR_PRESET;
+    if (process.env.LODESTAR_PRESET) delete process.env.LODESTAR_PRESET;
```
Comment (nflaig) [inline packages/params/test/e2e/overridePreset.test.ts]: what's the reason for also deleting `LODESTAR_PRESET=mainnet` now?

Comment (nazarhussain) [reply packages/params/test/e2e/overridePreset.test.ts]: We need to delete this env anyway because we override preset as hardcoded value in the `overridePresetOk.ts`

File: packages/logger/package.json (hunk 10 above)
Comment (nazarhussain) [reply packages/logger/package.json:64]: Nothing preset specific in these tests.

File: packages/params/test/e2e/overridePreset.test.ts (hunk 11 above)
Comment (nflaig) [reply packages/params/test/e2e/overridePreset.test.ts]: > We need to delete this env anyway because we override preset as hardcoded

I think the reason we have to delete it is this assertion
https://github.com/ChainSafe/lodestar/blob/c98a48a4fbbd2a4fd970d327683bf62161a93344/packages/params/test/e2e/overridePresetOk.ts#L20

The previous comment `// These commands can not run with minimal preset` does not make sense, any preset value fails this.

we should properly comment here why the preset env is removed

```
// `LODESTAR_PRESET` must not be set to correctly test if override works as expected
```

**Review by nflaig — DISMISSED** [review]

**Review by matthewkeil — APPROVED** [review]
LGTM!!

File: packages/beacon-node/package.json [hunk 12]
```diff
@@ -76,8 +76,8 @@
     "lint": "eslint --color --ext .ts src/ test/",
     "lint:fix": "yarn run lint --fix",
     "test": "yarn test:unit && yarn test:e2e",
-    "test:unit:minimal": "vitest --run --segfaultRetry 3 --dir test/unit/",
-    "test:unit:mainnet": "LODESTAR_PRESET=mainnet vitest --run --dir test/unit-mainnet",
+    "test:unit:minimal": "LODESTAR_PRESET=minimal vitest --run --segfaultRetry 3 --dir test/unit/",
+    "test:unit:mainnet": "LODESTAR_PRESET=mainnet vitest --run --segfaultRetry 3 --dir test/unit-mainnet",
```
Comment (matthewkeil) [inline packages/beacon-node/package.json:80]: I noticed this `segfaultRetry` was added.  what does that do?

File: packages/cli/test/scripts/e2e_test_env.ts [hunk 13]
```diff
@@ -14,6 +14,7 @@ const {forkConfig} = defineSimTestConfig({
   BELLATRIX_FORK_EPOCH: bellatrixForkEpoch,
   CAPELLA_FORK_EPOCH: capellaForkEpoch,
   runTillEpoch: Infinity,
+  initialNodes: 2,
```
Comment (matthewkeil) [inline packages/cli/test/scripts/e2e_test_env.ts:17]: i noticed this  was added.  what does that do?

File: scripts/run_e2e_env.sh [hunk 14]
```diff
@@ -1,18 +1,19 @@
 #!/bin/bash
 
+DIR=$(dirname "$0")
```
Comment (matthewkeil) [inline scripts/run_e2e_env.sh]: I’ve seen this done like the below before.
```sh
DIR="$(CDPATH= cd "$(dirname "${BASH_SOURCE[0]}")" && pwd)"
```

It ensures that the directory is obtained correctly even if the script is executed from a different directory or via a symbolic link

**Review by matthewkeil — CHANGES_REQUESTED** [review]
FWIW I am sorry I brought up the `LODESTAR_PRESET`.  Apologies!!  I didn’t realize the amount of confusion it would insert because of how we use the two for different tests interchangeably.   Thank you for the thoroughness in working through those challenges.

Just the two questions about the inserts and the update for getting the BASH dirname

File: packages/cli/test/scripts/e2e_test_env.ts (hunk 13 above)
Comment (nazarhussain) [reply packages/cli/test/scripts/e2e_test_env.ts:17]: It adjust the genesis delay for the e2e notes, without that default is a bit higher value, so tests take more time to wait for capella fork.

File: packages/beacon-node/package.json (hunk 12 above)
Comment (nazarhussain) [reply packages/beacon-node/package.json:80]: It was earlier added in `minimal` preset and missing in `mainnet`. It just check if tests fails with a segfault error, it tries to rerun it.

**Comment by nazarhussain** [comment]
> FWIW I am sorry I brought up the `LODESTAR_PRESET`. Apologies!!

@matthewkeil No problem at all. We had to clean up today or later anyway.

**Review by matthewkeil — DISMISSED** [review]
Sweet!! 🎸 I haven't run with the most recent changes but assuming it works.  Although I noticed that CI was not happy.  Not sure if that is related to these changes though....  A goal to shoot for is all green CI.  If I can help I will be glad to lend a hand.  Just point me in a direction and ill do my best

**Review by nflaig — DISMISSED** [review]

**Review by matthewkeil — DISMISSED** [review]
Reapproving but asked in discord about such a big downgrade to make sure we are not going to run into issues

File: .env.test [hunk 15]
```diff
@@ -0,0 +1,13 @@
+# We use these images during sim and e2e tests
+# TODO: Upgrade Geth once the Nethermind issue is reosolved else it's causing follorwing error
```
Comment (nflaig) [inline .env.test]:
```suggestion
# TODO: Upgrade Geth once the Nethermind issue is resolved else it's causing following error
```

**Review by nflaig — APPROVED** [review]

**Comment by wemeetagain** [comment]
:tada: This PR is included in [v1.16.0](https://github.com/ChainSafe/lodestar/releases/tag/v1.16.0) :tada:
